### PR TITLE
Reflect before invoking the callback

### PR DESCRIPTION
**Pull request by alshakero**
Fixes: https://github.com/Palindrom/JSONPatcherProxy/issues/21

@warpech can you review please? 

Reviewing while ignoring the whitespace saves you a minute or two (https://github.com/Palindrom/JSONPatcherProxy/compare/Reflect-earlier?expand=1&w=1)

### CODE REVIEW

**Comment by warpech** [comment]
I will review.

1. @alshakero I suppose the fix means that you fully agree with my claim that the callback should be called after the reflection?
2. @alshakero can you think of any risks involved with this change?
3. @miyconst may we include this in 2.3.1? I think we should

**Comment by alshakero** [comment]
> @alshakero I suppose the fix means that you fully agree with my claim that the callback should be called after the reflection?

I saw this as the desired behavior all along, it was surprising to me that this bug survived until now. JSONPatcherProxy IMO is meant to be a non-intrusive observer. Not an *intercepter* that gives the chance to throw errors/validate changes before reflection. I agree that reflection should happen before notification.

File: src/jsonpatcherproxy.js [hunk 1]
```diff
@@ -139,37 +142,42 @@ const JSONPatcherProxy = (function() {
...
       }
     } else {
+      reflectionResult = Reflect.set(target, key, newValue);
       instance.defaultCallback({
         op: 'add',
         path: destinationPropKey,
         value: newValue
       });
-      return Reflect.set(target, key, newValue);
+      return reflectionResult;
```
Comment (warpech) [inline src/jsonpatcherproxy.js]: `setTrap` has several calls to `Reflect.set`, `instance.defaultCallback`. The only thing that changes is the config object that you pass to `instance.defaultCallback`.

A more DRY thing to do would be to keep a single call at the end of `setTrap`:

```
const reflectionResult = Reflect.set(target, key, newValue);
instance.defaultCallback(patch);return reflectionResult;
```

**Comment by miyconst** [comment]
>@miyconst may we include this in 2.3.1? I think we should

Yes, we can and should.

**Pull request merged**